Lib/_collections_abc.py

```diff
@@ -962,7 +962,7 @@ def clear(self):
 
     def update(self, other=(), /, **kwds):
         ''' D.update([E, ]**F) -> None.  Update D from mapping/iterable E and F.
-            If E present and has a .keys() method, does:     for k in E: D[k] = E[k]
+            If E present and has a .keys() method, does:     for k in E.keys(): D[k] = E[k]
             If E present and lacks .keys() method, does:     for (k, v) in E: D[k] = v
             In either case, this is followed by: for k, v in F.items(): D[k] = v
```

Comment:
Here I would like it to be more specific:
```suggestion
            If E present and has keys attribute, does:     for k in E.keys(): D[k] = E[k]
            If E present and lacks keys attribute, does:     for (k, v) in E: D[k] = v
            In either case, this is followed by: for k, v in F.items(): D[k] = v
```
In the web docs the requirements are strict:
- `dict`: If a positional argument is given and it defines a `keys()` method, <...>. Otherwise, the positional argument **must be** an iterable object.
- `update`: **Accepts** either another object with a `keys()` method <...> or <...>.

Usage of words `must` and `accept` indicate that any object with non-callable `keys` is not a valid argument. But phrase `If E present and lacks .keys() method` does not eliminate cases with non-callable `keys`.
For example:
```python
class Object:
    keys = []
    def __iter__(self, /):
        return iter(self.keys)
```
This class clearly *lacks .keys() method* because `keys` is a list, but passing it to `dict.update` results in `TypeError: 'list' object is not callable`.

Reply:
This is fine IMO. It's much less important for docstrings to be completely precise than it is for the ReST documentation. The most important thing for the docstrings is for them to be concise and easily readable in a terminal or editor.

Reply:
I see. Any particular reason why is there no `a` before `.keys() method` in the second sentence? It is present in 3.12, but looks like it was removed prior this PR.
![image](https://github.com/user-attachments/assets/d92a2be9-94ba-4b6f-9bf6-1e71182dc57a)

Reply:
You're calling `help(dict.update)` there, which has a different docstring to `collections.abc.Mapping.update` (the one being edited here).

Reply:
Oh, my bad. In that case `dict.update` should also be updated as it works the same way (i.e. it calls `keys` attribute).

Reply:
Right, but the docstring for `dict.update` already appears to be correct, unless I'm misunderstanding.

Reply:
In this PR the first sentence was updated from `for k in E: D[k] = E[k]` to `for k in E.keys(): D[k] = E[k]` for MutableMapping. `dict.update` has the same behavior and its docstring has the same flaw, so it would be correct to fix it too. 
https://github.com/python/cpython/blob/2982bdb936f76518b29cf7de356eb5fafd22d112/Objects/dictobject.c#L4521-L4525

Consider:
```python
class Object:
    def keys(self):
        return [1, 2]
    def __iter__(self):
        return [3, 4]
    def __getitem__(self, item):
        match item:
            case 1: return 'one'
            case 2: return 'two'
            case 3: return 'three'
            case 4: return 'four'
        
        return item


print(dict(Object()))
d = {}
d.update(Object())
print(d)
```
Both times the result is `{1: 'one', 2: 'two'}`, i.e., both times `keys()` is called, the object is not iterated.

Reply:
Sorry, I see what you mean now. Yes, I suppose that could also be clarified.